temporian/beam/io.py

```diff
@@ -0,0 +1,299 @@
+# Note the double 2-items tuple (instead of a one 3-items tuple) to facilitate
+# Beam operations.
+StructuredRow = Tuple[
+    Tuple[BeamIndex, ...], Tuple[Timestamps, Tuple[np.generic, ...]]
+]
+
+# Unit of data for an event set used by all the operators implementations.
+#
+# Contains: the index+feature_idx, timestamps, and feature values.
+# The feature value can be None for event-set without features.
+BeamEventSet = Tuple[
+    Tuple[BeamIndexAndFeature, ...], Tuple[np.ndarray, Optional[np.ndarray]]
+]
+
+# From the point of view of the user, a "Beam event set" is a PCollection of
+# BeamEventSet.
+PColBeamEventSet = beam.PCollection[BeamEventSet]
+
+
+def _parse_csv_file(
+    file: beam.io.filesystem.FileMetadata,
+) -> Iterable[Dict[str, str]]:
+    """Parse a csv file into dictionary of key -> value."""
+
+    with beam.io.filesystems.FileSystems.open(file.path) as byte_stream:
+        string_stream = (x.decode("utf-8") for x in byte_stream)
+        for row in csv.DictReader(string_stream):
+            yield row
+
+
+@beam.ptransform_fn
+def read_csv_raw(pipe, file_pattern: str) -> beam.PCollection[Dict[str, str]]:
+    """Reads a file or set of csv files into a PCollection of key->values.
+
+    This format is similar to output of the official beam IO connectors:
+    https://beam.apache.org/documentation/io/connectors/
+
+    CSV values are always string, so the output of `read_csv_raw` is always
+    a dictionary of string to string. Use `to_event_set` (or better, use
+    `read_csv` instead of `read_csv_raw`) to cast values to the expected
+    pipeline input dtype.
+
+    Args:
+        pipe: A begin Beam pipe.
+        file_pattern: Path or path matching expression compatible with
+            `MatchFiles`.
+
+    Returns:
+        A PCollection of dictionary of key:value.
+    """
+
+    return (
+        pipe
+        | "List files" >> MatchFiles(file_pattern)
+        | "Parse file" >> beam.FlatMap(_parse_csv_file)
+    )
+
+
+def _cast_feature_value(value: Any, dtype: DType) -> np.generic:
+    """Convert a user feature value to the internal representation."""
+
+    np_type = tp_dtype_to_np_dtype(dtype)
+    return np_type(value)
+
+
+def _cast_index_value(value: Any, dtype: DType) -> BeamIndex:
+    """Convert a user index value to the internal representation."""
+
+    return _cast_feature_value(value, dtype).item()
+
+
+def _structure_fm(
+    row: Dict[str, str], schema: Schema, timestamp_key: str
+) -> StructuredRow:
+    """Transforms a dict of key:value to a StructuredRow."""
+
+    index_values = [
+        _cast_index_value(row[index.name], index.dtype)
+        for index in schema.indexes
+    ]
+    feature_values = [
+        _cast_feature_value(row[feature.name], feature.dtype)
+        for feature in schema.features
+    ]
+    timestamp = np.float64(row[timestamp_key])
+    return tuple(index_values), (timestamp, tuple(feature_values))
+
+
+class _MergeTimestampsDP(beam.DoFn):
```

Comment:
DP? a more expressive name + short docstring in the process method could help understand the merging logic